### ISSUE TITLE
⌘1 sometimes stops working in other apps

### ISSUE THREAD

**Issue by glyph** [zulip/zulip-desktop#317]
---
Please include: 
- `Operating System`:
    ```
    ProductName:	Mac OS X
    ProductVersion:	10.13
    BuildVersion:	17A405
    ```
- `Clear steps to reproduce the issue`:
    Run the application for a while.  Eventually, the ⌘1 hotkey starts getting captured regardless of where my keyboard focus is until I restart the application.

I am only logged in to one Zulip, but if I select settings, ⌘Tab away, and then hit ⌘1 in another app when in this state, I can see it switch back to that Zulip rather than the Settings window, so I assume that for those with multiple Zulips it would globally capture ⌘2, etc.

**Comment by akashnimare** [comment]
> ⌘1 sometimes stops working in other apps

@glyph can you define **sometimes** here? Does it happen all the time or in any specific case like when you switch to other apps using ⌘Tab?

**Comment by cedricium** [comment]
@zulipbot claim

**Comment by cedricium** [comment]
Hey @glyph, I'm going to be trying to fix this issue. In addition to defining what sometimes means, is there a specific application (when you <kbd>Cmd</kbd><kbd>Tab</kbd>) you're using that you see this issue happen with or is it multiple applications? I'm attempting to reproduce and haven't been able to so far. Thanks!

**Comment by timabbott** [comment]
@cedricium it seems likely the other application is irrelevant, and that the problem is that we're in some circumstance failing to unregister the hotkey from the OS...

**Comment by cedricium** [comment]
@timabbott that makes sense. I'm having trouble reproducing, am I understanding the issue correctly:

 1. In Zulip app, open the Settings page. Settings page is the active window.
 2. Switch to another application using <kbd>Cmd</kbd><kbd>Tab</kbd>. Different application is now focused.
 3. Use the <kbd>Cmd</kbd><kbd>1</kbd> shortcut in different application.
 4. Focus the Zulip app - active window is now the first sever (no longer the Settings page).

There's some sort of timing in between switching applications and using <kbd>Cmd</kbd><kbd>1</kbd> which causes the issue, but is the right way to go about it?

**Comment by akashnimare** [comment]
I'm guessing this is what @glyph experiencing -
* Open Zulip app, add a server
* Open setting page
* CMD+1 should switch to the first added server (this is expected)
* Now open other app, for an example Chrome
* Press CMD+1, which should open the first tab of Chrome but instead it focuses on Zulip app and opens the first server.

I can't reproduce the last step. I'm not sure if we have to unregister `CMD+1` when window loses focus, hidden etc since [electron-localshortcut](https://github.com/parro-it/electron-localshortcut) automatically unregister the shortcut for us.

**Comment by cedricium** [comment]
I still haven't been able to reproduce this issue, and looking at the code where the keyboard shortcuts for switching between servers gets set, I don't see anything that explicitly stands out which would cause this issue 😕

**Comment by akashnimare** [comment]
@glyph can you help us reproducing this issue. We've both tried to reproduce and read the code, and don't see how we can investigate further without more info 😕

**Comment by akashnimare** [comment]
@vbNETonIce can you reproduce this issue on Windows?

**Comment by penCsharpener** [comment]
no, never ran into this but I also don't really add new servers anymore (if that is relevant to the issues - just skimmed over this)

**Comment by akashnimare** [comment]
Hmm, thanks.
> I also don't really add new servers anymore 

Yeah, it's irrelevant. It could happen when you have only one server as well (since <kbd>CTRL+1/⌘1</kbd> gets assigned).

**Comment by glyph** [comment]
I am also noticing this with <kbd>⌘R</kbd> now, and it seems to happen almost every day now, which means part of my web-dev workflow is to quit zulip 3 or 4 times per day.

Anything I can do to make this more reproducible?

**Comment by glyph** [comment]
My OS is now
```
ProductName:	Mac OS X
ProductVersion:	10.13.1
BuildVersion:	17B48
```
so not fixed by Apple :)

**Comment by glyph** [comment]
I am not sure what the exact race condition is here, but https://www.npmjs.com/package/electron-localshortcut looks like a *horrible* hack that seems like a magnet for exactly this kind of bug. It is not appropriate for a mac app to register global hotkeys just to avoid populating the menu bar.  Please just put the stuff into a menu like the human interface guidelines recommend, and this global registration race won't even be possible.

**Comment by cedricium** [comment]
Hey @glyph, can verify the steps taken to reproduce this? Or if it's not always reproducible, just outline what happens when you use <kbd>⌘</kbd><kbd>1</kbd> / <kbd>⌘</kbd><kbd>R</kbd>. I didn't understand your original comment about when you opened this issue.

Would love to help get this fixed!

**Comment by timabbott** [comment]
@akashnimare what do you think about replacing that library?  I am partial to @glyph's argument here.

**Comment by glyph** [comment]
@cedricium I just launch Zulip and wait for it to steal my hotkeys ;).  What happens when I press <kbd>⌘1</kbd> with Zulip in the background is that it does exactly what <kbd>⌘1</kbd> does with Zulip in the foreground, whatever that would be.

**Comment by akashnimare** [comment]
I know `electron-localshortcut` has created problems in past but I guess we fixed most of the issues. I updated my macOS from `10.12` to `10.13` but couldn't repro this hotkey bug. @glyph 
just to make sure that we're on same page can you confirm that you're using the latest version which is `v1.5.0` via clicking on  `Menu > About Zulip` -

![image](https://user-images.githubusercontent.com/2263909/32727193-a5945cfc-c8a1-11e7-9bf2-1456dce6e085.png)

**Comment by glyph** [comment]
Yes, I'm using 1.5.0.

**Comment by glyph** [comment]
> I know electron-localshortcut has created problems in past but I guess we fixed most of the issues.

Only on your computer, apparently ;-).

The way electron-localshortcut works is *fundamentally* a bad idea.  Calling APIs like this for no reason is the sort of thing that gets you banned from the mac app store, flagged as malware, etc.  Global hotkeys should only be mapped if you are clearly calling this out to the user as global.

To help track this down, in case it's an interaction with something else, a list of applications I routinely run are:

- [Sip](http://sipapp.io/) (the old mac app store version)
- [OmniFocus](https://www.omnigroup.com/omnifocus/)
- [Emacs, mituharu's "mac port" version, installed from homebrew](https://github.com/railwaycat/homebrew-emacsmacport)
- [1password](https://1password.com/) (mac app store version)
- [Dropshare](https://getdropsha.re/)
- [Docker for Mac](https://www.docker.com/docker-mac)
- [Codepoints](https://itunes.apple.com/us/app/codepoints/id499161264?mt=12)
- [iStat Menus](https://bjango.com/mac/istatmenus/)
- Google Chrome
- [Moom](https://itunes.apple.com/gb/app/moom/id419330170?mt=12)
- [SizeUp](http://www.irradiatedsoftware.com/sizeup/)
- [Dash](https://kapeli.com/dash)
- [iTerm2](https://iterm2.com/)

There's nothing I can think of that I *do* with Zulip, or with any other app, that causes it to start grabbing the keyboard focus in the background; it just runs for a while and then I notice my command keys aren't doing what I expect.

**Comment by olbrich** [comment]
I see this pretty consistently with iTerm in particular.  Switching to Zulip (or quitting it) releases the cmd-1 key back to the system.   I can consistently get this to happen when I switch spaces.

**Comment by akashnimare** [comment]
Okay, so we need to remove `electron-localshortcut` library which is the root cause of this issue. I'm looking for ways to implement same shortcuts without the use of this module which should fix this issue.

**Comment by akashnimare** [comment]
I have pushed a fix in [remove-electron-localshortcut](https://github.com/zulip/zulip-electron/tree/remove-electron-localshortcut). I have removed `electron-localshortcut` dependency and now using electron's default [globalShortcut](https://github.com/electron/electron/blob/master/docs/api/global-shortcut.md) api to register the shortcuts. Can somebody cross check if it works as expected?

**Comment by glyph** [comment]
The problem with localshortcut is that it *uses* globalShortcut.  If anything this would be worse.

**Comment by akashnimare** [comment]
@glyph I have completely removed `electron-localshortcut` and `globalShortcut`. Now we only rely on menu for shortcuts - https://github.com/zulip/zulip-electron/commit/bcb8ffb55f7ef28c2b56bfee07322d1357f76458.

**Comment by akashnimare** [comment]
Just released [1.6.0-beta](https://github.com/zulip/zulip-electron/releases/tag/v1.6.0-beta). @glyph, @olbrich please update app (via selecting `Get beta updates` from setting option or download manually through this [link](https://github.com/zulip/zulip-electron/releases/download/v1.6.0-beta/Zulip-1.6.0-beta.dmg)) and see if it fixes this issue.

**Comment by glyph** [comment]
@akashnimare Fantastic. Thank you so much for the fix - and for fixing it the Right Way!

I got the auto-update notification and installed it, and despite using it for longer than the average time where this would trigger, with all my usual apps running, I have not had my hot keys get grabbed.  Seems to be working great!

It's hard to be *totally* sure since I don't know what I did that would reliably trigger it, but at this point I'd be confident closing this bug if I were you.  (Hopefully @olbrich can confirm.)
